hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java

```diff
@@ -461,25 +470,136 @@ public void run() {
         }
 
         private void rebalanceLeaderships() {
-            Map<RaftEndpoint, CPMember> members = getMembers();
+            Collection<CPMember> allMembers = getMembers().values();
+            Map<CPMember, Entry<Integer, Collection<CPGroupId>>> leadershipPriorityMap
+                    = getLeadershipPriorityMap(allMembers);
+
             Collection<CPGroupId> groupIds = getCpGroupIds();
+            List<CPGroupSummary> cpGroupSummaries
+                    = groupIds.stream()
+                    .map(this::getCpGroup)
+                    .collect(Collectors.toList());
+
+            // Transferring leaderships from the lowest priority CP members
+            Set<CPMember> priorityCPMembers
+                    = rebalanceLeadershipsByPriority(leadershipPriorityMap, cpGroupSummaries);
+
+            // If there are several CP members with the same priority level, we should rebalance leadership evenly among them
+            if (priorityCPMembers.size() > 1) {
+                rebalanceLeadershipsEvenly(priorityCPMembers, cpGroupSummaries, allMembers);
+            }
+        }
+
+        private Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>> getLeadershipPriorityMap(Collection<CPMember> members) {
+            Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>> leadershipsWithPriorities = new HashMap<>();
+            OperationService operationService = nodeEngine.getOperationService();
+            StringBuilder s = new StringBuilder("Current leadership claims:");
+            for (CPMember member : members) {
+                InvocationFuture<Map.Entry<Integer, Collection<CPGroupId>>> invocationFuture =
+                        operationService.invokeOnTarget(RaftService.SERVICE_NAME, new GetLeadedGroupsOp(),
+                                member.getAddress());
+                if (!invocationFuture.isCompletedExceptionally()) {
```

Comment:
how can we know if the future is completed exceptionally before it is not done? Shouldn't we check `isDone()` first?

Reply:
Thanks, @ahmetmircik for the comment!
Have refactored this part of the code with a `try-catch` block.